### PR TITLE
Use unsigned integers to represent addresses

### LINKED ISSUE #137: Segfault in persistent 4.6.0
I maintain the persistent package for Fedora.  I attempted to build version 4.6.0 for Fedora Rawhide.  The build was successful on all 64-bit architectures, but the test suite failed with segfaults on both 32-bit architectures: i686 and arm.  The segfaults are repeatable; I get them every time.  Running under valgrind shows no problems and the test suite completes successfully.  Running under gdb, however, I was able to gather some information.  The faulting test is `test_incrgc_simple (persistent.tests.test_picklecache.PythonPickleCacheTests)`.  The fault happens inside `PyObject_Repr`, on a bogus (garbage collected?) object.  This is what gdb has to say about the object:
```
(gdb) print *v
$1 = {ob_refcnt = -159554747, ob_type = 0x56703a34}
(gdb) print *$1->ob_type
$2 = {ob_base = {ob_base = {ob_refcnt = 0, ob_type = 0x0}, 
    ob_size = -158259432}, tp_name = 0x0, tp_basicsize = 0, tp_itemsize = 0, 
  tp_dealloc = 0x11, tp_vectorcall_offset = 1451167352, 
  tp_getattr = 0x566634d8, tp_setattr = 0x10, tp_as_async = 0x2e0, 
  tp_repr = 0x1, tp_as_number = 0xf7df84c0 <PyUnicode_Type>, 
  tp_as_sequence = 0x2be, tp_as_mapping = 0xffffffff, tp_hash = 0xa0d0de4, 
  tp_call = 0x0, tp_str = 0x70732041, tp_getattro = 0x61696365, 
  tp_setattro = 0x6f63206c, tp_as_buffer = 0x7274736e, tp_flags = 544498549, 
  tp_doc = 0x62617375 <error: Cannot access memory at address 0x62617375>, 
  tp_traverse = 0x7420656c, tp_clear = 0x6e61206f, 
  tp_richcompare = 0x61746f6e, tp_weaklistoffset = 1663067508, 
  tp_iter = 0x7373616c, tp_iternext = 0x6a626f20, tp_methods = 0x73746365, 
  tp_members = 0x200a0a2e, tp_getset = 0x46202020, tp_base = 0x6520726f, 
  tp_dict = <unknown at remote 0x706d6178>, tp_descr_get = 0x202c656c, 
  tp_descr_set = 0x70707573, tp_dictoffset = 543519599, tp_init = 0x68206577, 
  tp_alloc = 0x20657661, tp_new = 0x20656874, tp_free = 0x6c6c6f66, 
  tp_is_gc = 0x6e69776f, tp_bases = <unknown at remote 0x6c632067>, 
  tp_mro = <unknown at remote 0x65737361>, 
  tp_cache = <unknown at remote 0xa3a3a73>, 
  tp_subclasses = <unknown at remote 0x2020200a>, 
  tp_weaklist = <unknown at remote 0x63202020>, tp_del = 0x7373616c, 
  tp_version_tag = 1702057248, tp_finalize = 0x2e203a72, 
  tp_vectorcall = 0x20202e2e, tp_print = 0x62412023}
```

GDB's py-bt command produces this backtrace:
```
Traceback (most recent call first):
  <built-in method repr of module object at remote 0xf7986050>
  File "/usr/lib/python3.8/unittest/util.py", line 47, in safe_repr
    result = repr(obj)
  File "/usr/lib/python3.8/unittest/case.py", line 1321, in assertIsNone
    standardMsg = '%s is not None' % (safe_repr(obj),)
  File "/builddir/build/BUILD/persistent-4.6.0/persistent/tests/test_picklecache.py", line 344, in test_incrgc_simple
    self.assertIsNone(cache.get(oid))
  File "/usr/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/usr/lib/python3.8/unittest/case.py", line 1700, in run
  File "/usr/lib/python3.8/unittest/case.py", line 736, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.8/unittest/suite.py", line 378, in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.8/unittest/suite.py", line 378, in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.8/unittest/suite.py", line 378, in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.8/unittest/suite.py", line 378, in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.8/unittest/suite.py", line 378, in run
  File "/usr/lib/python3.8/unittest/suite.py", line 84, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python3.8/unittest/runner.py", line 432, in run
  File "/usr/lib/python3.8/unittest/main.py", line 783, in runTests
  File "/usr/lib/python3.8/unittest/main.py", line 101, in __init__
    self.runTests()
  File "/usr/lib/python3.8/site-packages/setuptools/command/test.py", line 512, in run_tests
  File "/usr/lib/python3.8/site-packages/setuptools/command/test.py", line 494, in run
  File "/usr/lib/python3.8/distutils/dist.py", line 985, in run_command
    cmd_obj.run()
  File "/usr/lib/python3.8/distutils/dist.py", line 966, in run_commands
    self.run_command(cmd)
  File "/usr/lib/python3.8/distutils/core.py", line 404, in setup
  File "/usr/lib/python3.8/site-packages/setuptools/__init__.py", line 144, in setup
    return distutils.core.setup(**attrs)
  File "setup.py", line 4432, in <module>
```
I looked at the diff between 4.5.1 and 4.6.0 and did not find the problem, but I did spot a couple of things that should probably be fixed.  In cPersistence.c, the `Py_DECREF(ts_module)` on line 1283 should rather be 2 lines up, I think.  In cPicklecache.c, the return on line 639 should probably be preceded by `Py_DECREF(dead_pers_obj);`, because that means that `PyDict_DelItem` did not decrement the refcount.  Please let me know what other information I can provide to help track this down.

### PR DESCRIPTION
**Pull request by jamadden**
Add change note.

Fixes #137

Supersedes and fixes #141 for committer agreement purposes.

### CODE REVIEW

File: CHANGES.rst [hunk 1]
```diff
@@ -4,7 +4,10 @@
 4.6.3 (unreleased)
 ------------------
 
-- Nothing changed yet.
+- Fix a crash in the test suite under a 32-bit CPython on certain
+  platforms. See `issue 137
+  <https://github.com/zopefoundation/persistent/issues/137>`_. Fix by
+  `Jerry James <https://github.com/jamesjer>`_.
```
Comment (mgedmin) [inline CHANGES.rst]: I got the impression that while the crash happened to be triggered in the test suite, it was actually a real bug and could just as well have crashed a real application?

Comment (mgedmin) [reply CHANGES.rst]: Also, s/certain platforms/32-bit platforms/ ?

Comment (jamadden) [reply CHANGES.rst]: It *could*, but it's unlikely. One would have to opt-in to using the pure-Python implementation on CPython (which is incredibly slow).

**Comment by jamadden** [comment]
Thanks!

**Pull request merged**